### PR TITLE
gpio-kbd-matrix: add drive column and direct access support

### PR DESCRIPTION
**Pull request by fabiobaltieri**
Add an option to select the drive mode of unselected columns and some optimization for the row/col access. I'll put together a document with the various configuration options for this driver once I'm done adding features. :-)

Still only tested on a dev board with jumper wires, seems to be working fine.

---

Add an option to drive inactive columns to inactive state rather than
high impedance. This is useful if the matrix has isolation diodes for
every key, as it allows the matrix to stabilize faster and the API for
changing the pin value is more efficient than the one to change the pin
configuration.

---

When the matrix is connected to consecutive pins on the same port, it's
possible to read the whole row or set the whole column in a single
operation. Fror the column, this is only possible if the matrix is
configured for driving unselected column, as there's no API to configure
multiple pins at the same time at the moment.

This is more efficient than checking the pins individually, and it's
particularly useful if the row or columns are driven from a GPIO port
expander.

Add some code to detect the condition and enable it automatically as
long as the hw configuration supports it.

### CODE REVIEW

File: dts/bindings/input/gpio-kbd-matrix.yaml [hunk 1]
```diff
@@ -39,5 +39,12 @@ properties:
...
-      The pins will be driven according to the GPIO_ACTIVE_HIGH or
-      GPIO_ACTIVE_LOW flags when selected, high impedance when not selected.
+      When unselected, this pin will be either driven to inactive state or
+      configured to high impedance (input) depending on whether the
+      col-drive-inactive property is set or not.
+
+  col-drive-inactive:
+    type: boolean
+    description: |
+      If enable, unselected column GPIOs will be driven to inactive state.
```
Comment (gmarull) [inline dts/bindings/input/gpio-kbd-matrix.yaml:49]:
```suggestion
      If enabled, unselected column GPIOs will be driven to inactive state.
```

Comment (fabiobaltieri) [reply dts/bindings/input/gpio-kbd-matrix.yaml]: Fixed

File: dts/bindings/input/gpio-kbd-matrix.yaml [hunk 2]
```diff
@@ -39,5 +39,12 @@ properties:
     required: true
     description: |
       GPIO for the keyboard matrix columns, supports up to 32 different GPIOs.
-      The pins will be driven according to the GPIO_ACTIVE_HIGH or
-      GPIO_ACTIVE_LOW flags when selected, high impedance when not selected.
+      When unselected, this pin will be either driven to inactive state or
+      configured to high impedance (input) depending on whether the
+      col-drive-inactive property is set or not.
```
Comment (keith-zephyr) [inline dts/bindings/input/gpio-kbd-matrix.yaml:44]:
```suggestion
      configured to high impedance (input) depending on the col-drive-inactive
      property.
```

File: drivers/input/input_gpio_kbd_matrix.c
Comment (keith-zephyr) [inline drivers/input/input_gpio_kbd_matrix.c:1]: Typo `FFor` in the commit message on https://github.com/zephyrproject-rtos/zephyr/pull/65167/commits/697aaf3be35a93dd7ea83118d49c90d16767908a

File: drivers/input/input_gpio_kbd_matrix.c [hunk 3]
```diff
@@ -102,10 +127,27 @@ static void gpio_kbd_matrix_set_detect_mode(const struct device *dev, bool enabl
 	}
 }
 
+static bool gpio_kbd_matrix_check_gpio_coherent(
```
Comment (keith-zephyr) [inline drivers/input/input_gpio_kbd_matrix.c:130]:
```suggestion
static bool gpio_kbd_matrix_is_gpio_coherent(
```

File: drivers/input/input_gpio_kbd_matrix.c [hunk 4]
```diff
@@ -102,10 +127,27 @@ static void gpio_kbd_matrix_set_detect_mode(const struct device *dev, bool enabl
...
+		const struct gpio_dt_spec *gpio, int gpio_count)
+{
+	const struct gpio_dt_spec *gpio0 = &gpio[0];
+
+	for (int i = 0; i < gpio_count; i++) {
```
Comment (keith-zephyr) [inline drivers/input/input_gpio_kbd_matrix.c:135]:
```suggestion
	for (int i = 1; i < gpio_count; i++) {
```

File: drivers/input/input_gpio_kbd_matrix.c [hunk 5]
```diff
@@ -50,6 +52,19 @@ static void gpio_kbd_matrix_drive_column(const struct device *dev, int col)
 		state = BIT(col);
 	}
 
+	if (data->direct_write) {
+		const struct gpio_dt_spec *gpio0 = &cfg->col_gpio[0];
+		gpio_port_pins_t gpio_mask;
+		gpio_port_value_t gpio_val;
+
+		gpio_mask = BIT_MASK(common->row_size) << gpio0->pin;
```
[the remark this answers is not in the document]
Comment (fabiobaltieri) [reply drivers/input/input_gpio_kbd_matrix.c:60]: nice! funny my test setup has 4 rows and 3 columns so this was somewhat working anyway, many thanks for catching it

File: drivers/input/input_gpio_kbd_matrix.c
Comment (fabiobaltieri) [reply drivers/input/input_gpio_kbd_matrix.c:1]: Ffixed

File: drivers/input/input_gpio_kbd_matrix.c (hunk 3 above)
Comment (fabiobaltieri) [reply drivers/input/input_gpio_kbd_matrix.c]: Done

File: drivers/input/input_gpio_kbd_matrix.c (hunk 4 above)
Comment (fabiobaltieri) [reply drivers/input/input_gpio_kbd_matrix.c]: Good point, done

File: drivers/input/input_gpio_kbd_matrix.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (fabiobaltieri) [reply drivers/input/input_gpio_kbd_matrix.c]: Fixed

File: dts/bindings/input/gpio-kbd-matrix.yaml (hunk 2 above)
Comment (fabiobaltieri) [reply dts/bindings/input/gpio-kbd-matrix.yaml]: Done

**Review by keith-zephyr — APPROVED** [review]

**Review by gmarull — APPROVED** [review]

**Pull request merged**